Add rectangular snap grid to osu! editor composer

bdach opened the pull request:
https://user-images.githubusercontent.com/20418176/133939345-60fa5f74-4459-4825-9391-014600efd89e.mp4

Appearance matching stable, also implements toggling grid size matching stable using the <kbd>G</kbd> key. The grid size is read from the beatmap file if stored. Contrary to stable however, the rectangular grid and the distance snap grid are mutually exclusive (for now? I just didn't see much point in having them both on at the same time)

Up to the grid toggling this was reasonably straightforward, but I wasn't sure where to put that part. The grid sizes available sort of rely on the semi-legacy 512x384 sizing of the playfield, so I made that stuff osu!-specific for now. May need to revisit that when work on the design tab starts.

The main grid component was written in a way that should be agnostic towards that (supports arbitrary start positions and spacings, not necessarily square ones even). Therefore, this also paves a potential way forward to allowing the grid to be rooted at the selected object, or custom grid spacings, if any of that is deemed as wanted.

The component's implementation is relatively ~~stupid~~ straightforward but doesn't seem to tank performance *too* much.

There was a fair few bits that I stole from the distance snap implementation but I didn't want to commonise that too much until the general direction is clear.

bdach committed:
Add rectangular position snap grid

bdach committed:
Add basic integration of rectangular grid to osu! composer

bdach committed:
Add test coverage for distance/rectangular grid exclusivity

bdach committed:
Implement grid size toggling matching stable

peppy commented:
Not too sure on having the grid *display* and *snap* be bound to the same button/key. I guess it's okay for now, but I believe there will be a sizeable group of mappers who want the grid to display but not snap (in stable the grid is always displayed).

A better behaviour may be to have the "grid snap" toggle implicitly turn the grid on, but having the grid cycle key have a final "off" state for those that want to disable the grid completely.

Also right now the grid defaults to off, and also in the highest density mode. Probably fine for the time being, but arguably not the best default when first starting on a new beatmap.

peppy committed:
Add key binding support for grid mode cycle

peppy reviewed (changes requested):
UX change proposed, haven't made the change myself as I wanted to get feedback first. Code looks fine otherwise.

bdach commented:
I'm ok with separating showing the rectangular grid and enabling snap for it, *but* I also think that the rectangular grid should hide when distance snap is enabled (else they will both show, which to me looks both visually busy and misleading). Stable didn't have this problem because stable's distance snap seems to have had little to no visual cues.

I'm also generally unsure of the usefulness of the grid if it doesn't snap (isn't it visually distracting?) but I suppose I generally wouldn't know as I've never seriously mapped anything.

Default density and starting state can be tweaked for sure.

peppy commented:
Seems like grid snap still occurs even after turning it off:

https://user-images.githubusercontent.com/191335/133986103-d4449e82-daa1-4012-afa9-3fb75400e9b0.mp4

Also the grid display level is reset every time you toggle grid snap:

https://user-images.githubusercontent.com/191335/133986158-d61cfad1-94cf-43f6-aba1-28619c2c1c6c.mp4

bdach committed:
Always show rectangular grid in osu! composer

bdach committed:
Fix rectangular grid snap being always active

bdach committed:
Adjust grid snap in line with new logic

bdach committed:
Start with largest grid size

bdach committed:
Store grid size back to beatmap on change

bdach commented:
Have made changes:

* Fixed the grid snap always being active (with test coverage).
* Changed starting grid size to largest (32).
* Changed the rectangular grid to show always (regardless of distance snap). It actually feels less cluttered now that the largest grid size is the default
  
  ![2021-09-20-203511_811x598_scrot](https://user-images.githubusercontent.com/20418176/134057945-0837703a-a341-4249-b841-e2163e7091b3.png)
  
  Check how you feel about that. I'll make it hide when distance snap is enabled on request.
* The grid snap choice is now stored back to the beatmap (not only read, as previously) and will persist across editor instances/restarts _if_ the beatmap is saved.

peppy committed:
Fix value not being loaded from beatmap in case of most dense grid setting

peppy commented on osu.Game.Rulesets.Osu/Edit/OsuHitObjectComposer.cs:
```diff
@@ -42,10 +42,12 @@ protected override DrawableRuleset<OsuHitObject> CreateDrawableRuleset(Ruleset r
         protected override IEnumerable<TernaryButton> CreateTernaryButtons() => base.CreateTernaryButtons().Concat(new[]
         {
-            new TernaryButton(distanceSnapToggle, "Distance Snap", () => new SpriteIcon { Icon = FontAwesome.Solid.Ruler })
+            new TernaryButton(distanceSnapToggle, "Distance Snap", () => new SpriteIcon { Icon = FontAwesome.Solid.Ruler }),
```
Not sure this should be the home for these buttons (since they aren't ternary) but it'll do for now.

peppy commented:
Pretty happy with this for the time being 👍 .

peppy committed:
Update test to match test beamap data
